lib/dictBuilder/cover.c

```diff
@@ -929,6 +929,65 @@ void COVER_best_finish(COVER_best_t *best, size_t compressedSize,
   }
 }
 
+static void COVER_selectDict(void* dict, size_t* dictBufferCapacity, const void* customDictContent, size_t dictContentSize,
+        const void* samplesBuffer, const size_t* samplesSizes, size_t nbTrainSamples, size_t nbSamples, ZDICT_cover_params_t params,
+        size_t* offsets, size_t* totalCompressedSize){
+
+  size_t largestDict = 0;
+  size_t largestCompressed = 0;
+  double defaultRegression = 1.01;
+
+  /* Initial dictionary size and compressed size */
+  {
+    *dictBufferCapacity = ZDICT_finalizeDictionary(
+      dict, *dictBufferCapacity, customDictContent, dictContentSize,
+      samplesBuffer, samplesSizes, (unsigned)nbTrainSamples, params.zParams);
+
+    if (ZDICT_isError(*dictBufferCapacity)) {
+      return;
+    }
+  }
+
+  *totalCompressedSize = COVER_checkTotalCompressedSize(params, samplesSizes,
+                                                       (const BYTE*)samplesBuffer, offsets,
+                                                       nbTrainSamples, nbSamples,
+                                                       (BYTE *const)dict, *dictBufferCapacity);
+
+  if(params.zParams.shrinkDict == 0){
+    return;
+  }
+
+  largestDict = *dictBufferCapacity;
+  largestCompressed = *totalCompressedSize;
+  *dictBufferCapacity = ZDICT_DICTSIZE_MIN;
+  dictContentSize = ZDICT_CONTENTSIZE_MIN;
+
+  /* Largest dict is initially at least ZDICT_DICTSIZE_MIN */
+  while (*dictBufferCapacity <= largestDict){
+    {
+      *dictBufferCapacity = ZDICT_finalizeDictionary(
+        dict, *dictBufferCapacity, customDictContent, dictContentSize,
+        samplesBuffer, samplesSizes, (unsigned)nbTrainSamples, params.zParams);
+
+      if (ZDICT_isError(*dictBufferCapacity)) {
+        return;
+      }
+    }
+
+    *totalCompressedSize = COVER_checkTotalCompressedSize(params, samplesSizes,
+                                                         (const BYTE*)samplesBuffer, offsets,
+                                                         nbTrainSamples, nbSamples,
+                                                         (BYTE *const)dict, *dictBufferCapacity);
+
+    if (*totalCompressedSize <= largestCompressed * defaultRegression){
+      break;
+    }
+    dictContentSize *= 2;
+    *dictBufferCapacity *= 2;
+  }
+  return;
```

Comment:
If we don't find any dictionaries better than the default one, this is buggy. Since we will have the wrong `dictBufferCapacity`, and `totalCompressedSize`.